Create basic model for Service Orders

lpichler opened the pull request:
https://trello.com/c/LVtwhl55/168-5-ssui-new-shopping-cart-model

cc @tinaafitz @gmcculloug @jrafanie

lpichler commented:
@miq-bot add-label wip

jrafanie commented on app/models/miq_request.rb:
```diff
@@ -5,6 +5,7 @@ class MiqRequest < ApplicationRecord
   belongs_to :destination,       :polymorphic => true
   belongs_to :requester,         :class_name  => "User"
   belongs_to :tenant
+  belongs_to :service_order
```
@gmcculloug do we have to worry about prior MiqRequests that aren't in a ServiceOrder?

jrafanie replied:
Note, I don't think so, just asking...

gmcculloug replied:
No, I do not think that is needed either.  Could always be a data migration at a later time if we change our minds.

gmcculloug commented on app/models/user.rb:
```diff
@@ -16,6 +16,7 @@ class User < ApplicationRecord
   has_many   :miq_widget_contents, :dependent => :destroy
   has_many   :miq_widget_sets, :as => :owner, :dependent => :destroy
   has_many   :miq_reports, :dependent => :nullify
+  has_many   :service_orders, :dependent => :nullify
```
I agree that when we delete a user we want to keep the service_orders around for things like reporting on order history per tenant, so `nullify` makes sense here.  However, we will lose any reference to the user in that object.  Does it make sense to add a column like "user_name", similar to the "requester_name" used in miq_requests?  (Note: Since we do have a link between the service_model and miq_requests we could also find the user that way.)

@dclarizio @jrafanie thoughts?

jrafanie replied:
Wouldn't we always get the list of service_orders via a user object?

Maybe, we'd want to see order history or something for the tenant/group, and not just user?  Is that what you're thinking?

dclarizio replied:
I can see wanting to run reports on order history either by tenant or globally. When we add costing, we will for sure not want to lose any order history and would certainly be nice to know who placed the orders, even if the user records have been removed.

lpichler commented:
@jrafanie @gmcculloug @dclarizio I added storing user_name to miq_request and I also added default value for user (current_user)

jrafanie commented on db/migrate/20160209130938_add_service_order_and_user_name_to_miq_request.rb:
```diff
@@ -0,0 +1,6 @@
+class AddServiceOrderAndUserNameToMiqRequest < ActiveRecord::Migration
+  def change
+    add_column :miq_requests, :service_order_id, :bigint
+    add_column :miq_requests, :user_name, :string
```
@lpichler, we already have userid column and requester_name in miq_requests, I believe @gmcculloug is suggesting adding user_name to service_orders.

gmcculloug replied:
Correct, we want to capture the user that created the Service Order so we have some reference if that user object gets deleted.  MiqRequest already has this data which I recall was added to avoid additional lookups when viewing requests but also works for this purpose as well.

jrafanie commented on app/models/service_order.rb:
```diff
@@ -0,0 +1,15 @@
+  default_value_for :user_id, User.current_user.id
+
+  def assign_user_name_to(miq_request)
+    miq_request.user_name = user.name
```
Since we want to capture the user_name of the user the service order is for, we should probably do something like `self.user_name = User.current_user.id` in a before create and not in default_value_for.

jrafanie replied:
@gmcculloug does it matter if we use the `User.current_user` for the service order?  Is it possible to create a service order for someone else, such as sharing a wish list with someone?

gmcculloug replied:
We have not worked out the details of how a wish-list work would just yet so I would not code for that now.  I would say when we create the ServiceOrder we will be passing the user object to the create call and we can use that object to capture the `user_name` as well as the `tenant_id` at that time.  Then the caller will be responsible for passing the user object with the proper identify.

lpichler commented:
thanks @jrafanie @gmcculloug I added `user_name` to `service_order` (instead of `miq_request`) with populating `user_name` on `before_create` and removed assigning current user.

lpichler commented:
@miq-bot remove-label wip

gmcculloug commented on app/models/service_order.rb:
```diff
@@ -0,0 +1,15 @@
+  before_create :assign_user_name
+
+  def assign_user_name
+    self.user_name = user.name unless user.nil?
```
Use `self.user_name = user.try(:name)`

lpichler replied:
@gmcculloug :+1:

gmcculloug commented:
:+1: Looks good.

jrafanie commented:
Nice @lpichler.  Thanks @gmcculloug @dclarizio for the feedback.
